### ISSUE TITLE
JUnit reporter: Output XML doesn't contain any details regarding failures on test "step"s

### ISSUE THREAD

**Issue by magurotuna** [denoland/deno#22795]
## Issue

We can write nested test cases using `t.step` (where `t` is [`TestContext`]).
When we run `deno test --reporter=junit` for a test file containing nested test cases that fail, the output XML will not contain useful information such as the place where it failed or how it failed (e.g. assertion messages).

[`TestContext`]: https://deno.land/api@v1.41.2?s=Deno.TestContext

### Deno

Say we have the following test file:

```ts
import { assertEquals } from "https://deno.land/std@0.217.0/assert/mod.ts";

Deno.test({
  name: "parent",
  fn: async (t) => {
    await t.step("step1", () => {
      assertEquals(42, 5); // should fail
    });
    await t.step("step2", () => {
      assertEquals(42, 42); // should succeed
    });
  },
});
```

Running `deno test --reporter=junit` gets us:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="deno test" tests="1" failures="1" errors="0" time="0.003">
    <testsuite name="file:///private/tmp/tmpmar8/main_test.ts" tests="1" disabled="0" errors="0" failures="1">
        <testcase name="parent" time="0.002" filename="/private/tmp/tmpmar8/main_test.ts" line="3" col="6">
            <properties>
                <property name="step[failure]" value="step1"/>
                <property name="step[passed]" value="step2"/>
            </properties>
            <failure message="1 test step failed."/>
        </testcase>
    </testsuite>
</testsuites>
```

We can tell the fact that `step1` failed and `step2` succeeded, but no other information is available. In particular we would like to get assertion error messages like `Actual 42, Got 5` or something.

### Other test runners

Take a look at vitest for comparison. 

Here's a input to provide to vitest. Note that the structure is the same as we did above for `deno test`.

```ts
import { describe, it } from "vitest";

describe("parent", () => {
  it("step1", ({ expect }) => {
    expect(42).toBe(5); // should fail
  });
  it("step2", ({ expect }) => {
    expect(42).toBe(42); // should succeed
  });
});
```

Running vitest with its JUnit reporter enabled generates the following output:

```xml
<?xml version="1.0" encoding="UTF-8" ?>
<testsuites name="vitest tests" tests="2" failures="1" errors="0" time="0.121">
    <testsuite name="vitest.test.ts" timestamp="2024-03-08T12:39:17.635Z" hostname="Yusukes-MacBook-Air.local" tests="2" failures="1" errors="0" skipped="0" time="0.004">
        <testcase classname="vitest.test.ts" name="parent &gt; step1" time="0.003">
            <failure message="expected 42 to be 5 // Object.is equality" type="AssertionError">
AssertionError: expected 42 to be 5 // Object.is equality
 ❯ vitest.test.ts:5:16
            </failure>
        </testcase>
        <testcase classname="vitest.test.ts" name="parent &gt; step2" time="0.001">
        </testcase>
    </testsuite>
</testsuites>
```

This shows that the failed test step gives an assertion error message and where it failed, which is pretty nice.

My proposal here is that we improve Deno's JUnit reporter so it can output as much information as vitest.

## Version

Version: Deno 1.41.2

**Comment by bartlomieju** [comment]
@magurotuna is this by any chance the same as https://github.com/denoland/deno/issues/20154?

**Comment by magurotuna** [comment]
@bartlomieju I didn't find that one when opening mine but yes, that looks pretty related to the issue I was raising in here, though #20154 proposes a more flexible approach where it would provide three options on whether to output parent (i.e. non-leaf) test cases.
I think that https://github.com/denoland/deno/pull/20273#discussion_r1307558757 described the reason very well as to why these options are needed to meet all the possible scenarios users may encounter depending on how they structure test hierarchy. But I also think that if we can accept that one failure happening in a nested test step will end up being counted twice (one for the test step itself, and the other for its parent), then we can make the implementation very simple and don't have to provide options - my fix in #22797 is based on this idea. I'll write considerations and design choices in a bit more detail in #22797
